Update Identity Components in Blazor project template

halter73 opened the pull request:
## Description

This PR addresses a large amount of feedback from #50722 which was merged before they could all be addressed to unblock Accessibility Testing effort. The primary impacts are:

#### Runtime changes
- Public API change to make `AddComponentRenderMode`'s `renderMode` param nullable to support disabling interactivity on a per-page basis with the help of `@rendermode="null"` (effectively).
  - **IMPORTANT:** This will need follow up in the Razor Compiler. See https://github.com/dotnet/razor/issues/9343
  - API Proposal issue: #51170
  - This is a e necessary to support the changes to add global interactivity to Identity components @SteveSandersonMS made in #50920 and have now been included in this PR.
- [Add antiforgery token to forms rendered interactively on the server](https://github.com/dotnet/aspnetcore/pull/51134/commits/425bd12a5a442711271761df91b580c785f38fa5)
  - This bug fix is necessary to make the logout button work without throwing antiforgery errors when it is rendered interactively on the server.

#### Template changes

- Fix compilation error due to missing `using` in `Program.cs` when the individual auth option is selected with no interactivity.
- Add support for global (`--all-interactive`) instead of just per-page interactivity to the new Identity components.
- Fix "Apply Migrations" link on the `DatabaseErrorPage` by calling `UseMigrationsEndPoint()` when necessary. 
- Add support for non-root base paths to the new Identity components.
- Improve folder layout by putting most of the additional auth and Identity related files in the same /Account folder.
- Use the new `IEmailSender<ApplicationUser>` API instead of `IEmailSender` for easier customization of emails.
- Remove usage of `IHttpContextAccessor` from the template because that is generally regarded as bad practice due to the unnecessary reliance on `AsyncLocal`.
- Remove underscore (`_`) from private field names.
- Reduce usage of `null!` and `default!`.
- Use normal `<button>` for logout link in nav bar rather than `<a onclick="document.getElementById('logout-form').submit();">`, and remove separate `LogoutForm.razor`.

## Customer Impact

This fixes several bugs in the Blazor project template when choosing the individual auth option and makes several runtime fixes that will be beneficial to any global interactive Blazor application that needs to include some components that must always be statically rendered.

## Regression?

- [ ] Yes
- [x] No

## Risk

- [ ] High
- [ ] Medium
- [x] Low

Obviously, we would rather not make such a large change after RC2. Particularly when it's a change that touches public API. Fortunately, the runtime changes are very small, and only to parts of the runtime that were last updated in RC2 (see #50181 and #50946).

The vast majority of the changes in the PR only affect the Blazor project template when the non-default individual auth option is selected. This was merged very late in RC2 (#50722) with the expectation that we would make major changes prior to GA.

## Verification

- [x] Manual (required)
- [x] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

SteveSandersonMS committed:
Add auth with global interactivity

halter73 committed:
Address API review feedback

halter73 committed:
Make AddComponentRenderMode param nullable

halter73 committed:
Update Can~not~AddNullComponentRenderMode test

halter73 committed:
Fix NavMenu.razor output

- Move "/Logout" endpoint
- Make App.razor more consistent between modes

halter73 committed:
Add antiforgery token to forms rendered interactively on the server

halter73 committed:
Simplify Email.razor

javiercn commented:
@halter73 took a look at the changes.

Overall looks good, the main questions that I have are:
* The use of `required` which is "controversial"
* Whether the identity pages should live inside an `Identity` subfolder. It might feel unnecessary but its better for the longer term, as it enables people to use `_Import.razor` files that target only identity and nothing more.
* Should we have an IdentityPage base component?
  * There's a lot of repetition on the pages, injecting a bunch of services, getting the HttpContext, etc. all of which can be centralized into a base class.
  * We can then use an `@inherits IdentityPage` on an `_Imports.razor` file to apply the base to all the Identity pages in one go.

halter73 commented:
> * The use of `required` which is "controversial"

There was earlier discussion about this in the PR that originally added the Identity components to the project template when I added it the old way without `required` and instead used `= default!`. https://github.com/dotnet/aspnetcore/pull/50722#discussion_r1334590203.

I think

```csharp
[CascadingParameter]
public required HttpContext HttpContext { get; set; }
```

looks a lot nicer than

```csharp
[CascadingParameter]
public HttpContext HttpContext { get; set; } = default!;
```

And we even got some [community agreement.](https://github.com/dotnet/aspnetcore/pull/50722#issuecomment-1732809229) Although they also noted that this goes directly against [our guidance](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/?view=aspnetcore-7.0#component-parameters):

> Don't use the [required modifier](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/required) or [init accessor](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/init) on component parameter properties. Components are usually instantiated and assigned parameter values using [reflection](https://learn.microsoft.com/en-us/dotnet/csharp/advanced-topics/reflection-and-attributes/), which bypasses the guarantees that init and required are designed to make.

@MackinnonBuck also brought up a good point about how it could prevent code-generated parameter setting from working in the future if we ever decided to do that. We're only using `required` in situations like static server rendering where we theoretically could ensure the parameters like the `HttpContext` `[CascadingParameter]` already exist during component construction meaning it's possible we could make code generation work, but this would require initializing the parameters earlier than before which might break backwards compatibility to some extent.

I'm fine with whatever we decide. We're still using `private ApplicationUser user = default!;` in a bunch of places elsewhere, so it's not like these `required` modifiers is allowing us to completely remove `default!` from the template. I'd like to hear from @SteveSandersonMS before removing the `required`'s though, since he suggested it.

> * Whether the identity pages should live inside an `Identity` subfolder. It might feel unnecessary but its better for the longer term, as it enables people to use `_Import.razor` files that target only identity and nothing more.

The `Account` folder already serves this purpose. The [identity-specific _Imports.razor file](https://github.com/dotnet/aspnetcore/blob/c5663da2813ed738787e65e158b15577c6037434/src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp/Components/Account/Pages/_Imports.razor) you propopse already exists too, thanks to @SteveSandersonMS's changes in #50920.

> * Should we have an IdentityPage base component?

[Here's](https://github.com/dotnet/aspnetcore/pull/50722#discussion_r1333980553) the discussion from when you suggested it in the original PR. @SteveSandersonMS mentioined:

> While having a base class would work, it is a bit of a constraining solution (people may have their own other base classes they want to use). I'd be in favour of just adding `[CascadingParameter] private required HttpContext HttpContext { get; set; }` wherever it's needed (and I know it's in about 15 places).

It was actually less than 15 places, but it was pretty close. I agree there is a lot of repetition, but there are also quite a few very simple components like [Lockout.razor](https://github.com/dotnet/aspnetcore/blob/51f80c73e5a9bfb21962db8daa9bb174b9af5942/src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp/Components/Pages/Account/Lockout.razor) that gain nothing from a base component.

I'm not sure I buy that a base class is all that constraining. If people have their own base component, couldn't they make Identity base component inherit from their own? But I do think avoiding inheritance where possible is generally a good idea because it shows there's nothing up our sleeves so to speak. All of the services and parameters we use, we declare directly on the page.

MackinnonBuck commented on src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp/Components/Account/Pages/Manage/Email.razor:
```diff
@@ -0,0 +1,116 @@
+        email = await UserManager.GetEmailAsync(user);
+        isEmailConfirmed = await UserManager.IsEmailConfirmedAsync(user);
+
+        Input.NewEmail ??= email!;
```
Is this correct anymore, now that `Input.NewEmail` has a default value of `""`? I believe this piece of logic made it so that the input field's default value was the user's existing email.

MackinnonBuck commented on src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp/Components/Account/Pages/ResetPassword.razor:
```diff
@@ -72,15 +70,13 @@
         if (result.Succeeded)
         {
-            RedirectManager.RedirectTo("/Account/ResetPasswordConfirmation");
-            return;
```
It's kind of subtle that `RedirectTo()` never returns, making the `return` statement redundant. From reading the code, it might be hard to tell at first that each of these `if` statements acts as a kind of implicit guard clause. But I don't feel a strong way about this so if you think it's fine to remove the `return`s then I won't fight it! 🙂

halter73 committed:
Fix Email.razor's default NewEmail

halter73 committed:
Add LogoutForm.razor back

SteveSandersonMS commented on src/Components/Components/src/Rendering/RenderTreeBuilder.cs:
```diff
@@ -682,10 +682,8 @@ public void AddComponentReferenceCapture(int sequence, Action<object> componentR
-    public void AddComponentRenderMode(IComponentRenderMode renderMode)
+    public void AddComponentRenderMode(IComponentRenderMode? renderMode)
     {
-        ArgumentNullException.ThrowIfNull(renderMode);
```
I think you need to `return` explicitly if it's null, so it truly becomes a no-op. Otherwise you're setting `HasCallerSpecifiedRenderMode` even though there isn't a nonnull mode, and that will produce an actually different behavior than we had before.

halter73 committed:
External login improvements

halter73 committed:
Don't really add a null render mode to the tree

- This avoids making HasCallerSpecifiedRenderMode true in the @rendermode="null" case

halter73 committed:
Revert "Add LogoutForm.razor back"

- There's no good place for it currently when it's referenced by the client project

This reverts commit fac57d10e2f1bab452c0e4de4453fb6c4a8778b9.

halter73 commented:
You can review the project's generated with these template changes at https://github.com/halter73/BlazorWebTemplateMatrix#readme. You can just install the latest 8.0 SDK from the https://github.com/dotnet/installers repo (8.0.100-rtm.23506.1 currently) and reference that in the `global.json` to quickly try any of the variations. The -ai (i.e. `--all-interactivity`) versions compile but will have nullability warnings and throw `ArgumentNullException`s until we get the runtime fixes in this PR and the razor compiler `@rendermode` nullablity update tracked by #9343 into the SDK.

danroth27 commented:
@JeremyLikness @halter73 Are we sure we want to use "auth" in the template? The problem with "auth" is that it's an abbreviation for both authentication & authorization, which makes it ambiguous and it's also pretty informal.

Some alternatives:

- Authenticated.razor
- LoginRequired.razor
- Restricted.razor
- ...

danroth27 commented:
Also, the existing web templates in .NET 7 all put the login/logout and user account links in the top row on the upper right. Is there a reason that we've diverged from that design in the new Blazor Web App template?

javiercn commented on src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp/Components/Account/IdentityRedirectManager.cs:
```diff
@@ -0,0 +1,58 @@
+
+        // On the server, NavigateTo throws NavigationException on the server which is handled by the framework as a redirect
+        // before the InvalidOperationException is thrown.
+        navigationManager.NavigateTo(uri);
+        throw new InvalidOperationException($"{nameof(IdentityRedirectManager)} must be called from the server.");
```
Not for this PR, but in wasm/server `navigationManager.NavigateTo(uri);` continues executing, which is a visible behavior. /cc: @SteveSandersonMS

halter73 committed:
private HttpContext HttpContext { get; set; } = default!;

halter73 committed:
Merge branch 'release/8.0'

halter73 committed:
Test more "dotnet new blazor" project templates

- Add tests for "--interactivity WebAssembly"
- Add tests for "--interactivity Auto"
- Add tests for "--all-interactive"
- Add tests for "--empty"
- Add tests for "--auth Individual"
- Add tests for "--use-local-db"
- Update="Data\app.db"
- Fix reference to dotnet-ef.dll in PrepareForTest.targets

halter73 committed:
Test Auth.razor

halter73 committed:
Fix app.css color replacement

halter73 committed:
Merge branch 'release/8.0'

halter73 committed:
clean up

Kumima commented:
@halter73 I've noticed that the logout form has been changed into a standard form. But I did not find the endpoint definition of Account/Logout. Thre will be no handler for the logout form. If I get something wrong?

los93sol commented:
I routinely put an _Imports.razor at the root of my projects with @attribute [Authorize] to enforce authorization on all pages, but because @attribute [AllowAnonymous] is not explicitly defined on the razor pages in Pages/Account that ends up breaking the pages and requires me to go specify it manually.
